diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9a8b5c0..8946225 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,6 +46,8 @@ option ( LUA_USE_ASSERT "Turn on assertions for the whole LuaJIT VM. This signif
 option ( LUAJIT_CPU_SSE2 "Disable SSE2." OFF )
 option ( LUAJIT_CPU_NOCMOV "Disable NOCMOV." OFF )
 
+option ( LUAJIT_SHARED "Build shared library instead of static linked one." OFF )
+
 # Tunable variables
 set ( LUAI_MAXSTACK 65500 CACHE NUMBER "Max. # of stack slots for a thread (<64K)." )
 set ( LUAI_MAXCSTACK 8000 CACHE NUMBER "Max. # of stack slots for a C func (<10K)." )
@@ -72,7 +74,10 @@ set ( LUA_CPATH_DEFAULT "./?${LUA_MODULE_SUFFIX};${LUA_DIR}${LUA_CDIR}/?${LUA_MO
 
 if ( WIN32 AND NOT CYGWIN )
   # Windows systems
-  add_definitions ( -DLUA_BUILD_AS_DLL -DLUAJIT_OS=LUAJIT_OS_WINDOWS)
+  add_definitions ( -DLUAJIT_OS=LUAJIT_OS_WINDOWS)
+  if (LUAJIT_SHARED)
+    add_definitions ( -DLUA_BUILD_AS_DLL)
+  endif()
   option ( LUA_BUILD_WLUA "Build wluajit interpreter for no-console applications." ON )
   set ( LJVM_MODE peobj )
   # Paths (Double escapes needed)
@@ -303,7 +308,11 @@ set ( LJCORE_C
   ${LJLIB_C}
 )
 
-add_library ( liblua SHARED ${LJCORE_C} ${DEPS} )
+if ( LUAJIT_SHARED)
+  add_library ( liblua SHARED ${LJCORE_C} ${DEPS} )
+else()
+  add_library ( liblua ${LJCORE_C} ${DEPS} )
+endif()
 target_link_libraries ( liblua ${LIBS} )
 set_target_properties ( liblua PROPERTIES OUTPUT_NAME lua CLEAN_DIRECT_OUTPUT 1 )
 
@@ -318,11 +327,14 @@ if ( LUA_BUILD_WLUA )
   install_executable ( wluajit )
 endif ()
 
+export(TARGETS ${PROJECT_NAME}
+       NAMESPACE ${PROJECT_NAME}::
+       FILE "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}-config.cmake")
+
 ## INSTALL
 install_executable( luajit )
 install_library( liblua )
 install_header( src/lua.h src/luajit.h src/lua.hpp ${CMAKE_CURRENT_BINARY_DIR}/luaconf.h src/lualib.h src/lauxlib.h )
-
-install_doc ( doc/ )
-install_foo ( etc/ )
-install_data ( COPYRIGHT README.md )
+install(EXPORT ${PROJECT_NAME}-config
+        DESTINATION "${CMAKE_INSTALL_LIBDIR}/${PROJECT_NAME}/cmake"
+        NAMESPACE ${PROJECT_NAME}::)
diff --git a/cmake/dist.cmake b/cmake/dist.cmake
index 310ef94..7c5130b 100644
--- a/cmake/dist.cmake
+++ b/cmake/dist.cmake
@@ -160,6 +160,7 @@ macro ( install_library )
                               SOVERSION ${DIST_VERSION} )
     endif ()
     install ( TARGETS ${_file}
+              EXPORT luajit-config
               RUNTIME DESTINATION ${INSTALL_BIN} COMPONENT Runtime
               LIBRARY DESTINATION ${INSTALL_LIB} COMPONENT Runtime 
               ARCHIVE DESTINATION ${INSTALL_LIB} COMPONENT Library )
